gene/database/postgresql.py

```diff
@@ -0,0 +1,867 @@
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Set
+import tempfile
+from datetime import datetime
+
+import psycopg
+from psycopg.errors import UndefinedTable, UniqueViolation
+import requests
+
+from gene.database import AbstractDatabase, DatabaseException, \
+    DatabaseReadException, DatabaseWriteException
+from gene.schemas import SourceMeta, SourceName
+
+
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
+
+
+class PostgresDatabase(AbstractDatabase):
+    """Database class employing PostgreSQL."""
+
+    def __init__(self, db_url: Optional[str] = None, **db_args) -> None:
+        """Initialize Postgres connection.
+
+        :param db_url: libpq compliant database connection URI
+        :param **db_args: see below
+
+        :Keyword Arguments:
+            * user: Postgres username
+            * password: Postgres password (optional or blank if unneeded)
+            * db_name: name of database to connect to
+
+        :raise DatabaseInitializationException: if initial setup fails
+        """
+        if db_url:
+            conninfo = db_url
+        elif "GENE_NORM_DB_URL" in os.environ:
+            conninfo = os.environ["GENE_NORM_DB_URL"]
+        else:
+            user = db_args.get("user", "postgres")
+            password = db_args.get("password", "")
+            db_name = db_args.get("db_name", "gene_normalizer")
+            if password:
+                conninfo = f"dbname={db_name} user={user} password={password}"
+            else:
+                conninfo = f"dbname={db_name} user={user}"
+
+        self.conn = psycopg.connect(conninfo)
+        self.initialize_db()
+        self._cached_sources = {}
+
+        atexit.register(self.close_connection)
+
+    def list_tables(self) -> List[str]:
+        """Return names of tables in database.
+
+        :return: Table names in database
+        """
+        with self.conn.cursor() as cur:
+            cur.execute(
+                """SELECT table_name FROM information_schema.tables
+                WHERE table_schema = 'public'
+                AND table_type = 'BASE TABLE';
+                """
+            )
+            tables = cur.fetchall()
+        return [t[0] for t in tables]
+
+    def drop_db(self) -> None:
+        """Perform complete teardown of DB. Useful for quickly resetting all data or
+        reconstructing after apparent schema error.
+        """
+        drop_query = """
+        DROP MATERIALIZED VIEW IF EXISTS record_lookup_view;
+        DROP TABLE IF EXISTS
+            gene_associations,
+            gene_symbols,
+            gene_previous_symbols,
+            gene_aliases,
+            gene_xrefs,
+            gene_concepts,
+            gene_merged,
+            gene_sources;
+        """
+        with self.conn.cursor() as cur:
+            cur.execute(drop_query)
+            self.conn.commit()
+        logger.info("Dropped all existing gene normalizer tables.")
+
+    def initialize_db(self) -> None:
```

Comment:
It might be nicer to have this part be a separate sql file. It would make it easier to read the commands as well

Reply:
I took the bigger/more static queries and put them into their own files. This module is still a little long but hopefully it's closer to readable now